Sources/SwiftGraphQLCodegen/Generator/Field.swift

```diff
@@ -1,306 +1,328 @@
 import Foundation
+import GraphQLAST
 
 /*
- Each field decoder contains a selection part that is responsible for
+ We use functions - selections - to construct a query.
+
+ Each selection function contains a selection part that is responsible for
  telling the Selection about its existance and a decoder part that
  checks for the result and returns it as a function value.
- 
+
  The last part of the function is a mock value which we use as a placeholder
  of the return value on the first run when we collect the selection.
  */
 
-extension GraphQLCodegen {
-    // MARK: - Field Selection
-    
-    /// Generates a SwiftGraphQL field.
-    func generateField(_ field: GraphQL.Field) throws -> [String] {
-        var lines = [String]()
-        
-        // Documentation.
-        if let docs = generateFieldDoc(for: field) {
-            lines.append(docs)
-        }
-        if let deprecationDocs = generateFieldDeprecationDoc(for: field) {
-            lines.append(deprecationDocs)
-        }
-        
-        // Field method.
-        lines.append("func \(try generateFnDefinition(for: field)) throws -> \(try generateReturnType(for: field.type)) {")
-        lines.append("    /* Selection */")
-        lines.append(contentsOf: generateFieldSelection(for: field).indent(by: 4))
-        lines.append("    self.select(field)")
-        lines.append("")
-        lines.append("    /* Decoder */")
-        lines.append("    switch self.response {")
-        lines.append("    case .decoding(let data):")
-        lines.append(contentsOf: generateDecoder(for: field).indent(by: 8))
-        lines.append("    case .mocking:")
-        lines.append("        return \(try generateMockData(for: field.type))")
-        lines.append("    }")
-        lines.append("}")
-        
-        return lines
-    }
-    
-    // MARK: - Documentation
-    
-    /// Generates field documentation.
-    private func generateFieldDoc(for field: GraphQL.Field) -> String? {
-        field.description.map { "/// \($0)" }
+// MARK: - Function Definition
+
+/*
+ This section contains functions that we use to generate the selection
+ function itself.
+ */
+
+extension Collection where Element == Field {
+    /// Returns the functions that represent selection for a given type.
+    func selection(scalars: ScalarMap) throws -> String {
+        try map { try $0.selection(scalars: scalars) }.joined(separator: "\n")
     }
-    
-    /// Generates deprecation documentation.
-    private func generateFieldDeprecationDoc(for field: GraphQL.Field) -> String? {
-        field.isDeprecated ? "@available(*, deprecated, message: \"\(field.deprecationReason ?? "")\")" : nil
+}
+
+extension Field {
+    /// Returns the function that may be used to create selection using SwiftGraphQL.
+    func selection(scalars: ScalarMap) throws -> String {
+        """
+        \(docs)
+        \(availability)
```

Comment:
I think you were gonna remove availability anyway but here if it's not there the docs don't attach to the field. They still work surprisingly.

Example:
<img width="700" alt="image" src="https://user-images.githubusercontent.com/17001151/107390062-9ab5bc80-6af7-11eb-99f6-e2ac97d06f22.png">